Add ability for drivers to set static routes

tomdee opened the pull request:
As suggested by @mrjana  and @mavenugo here's a rough outline for the API changes required to allow drivers to set static routes in the sandbox. 

Can you comment on this approach? If you think you could accept this PR I'll write the implementation and UTs.

I also expect I'll need to change the remote driver once @squaremo 's changes go in.

squaremo commented:
This seems entirely reasonable as a proposal. There will need to be some more (un)marshalling code for the remote driver, but all the bits are there to do it.

mavenugo commented:
@tomdee apologies on the delay. overall the structure looks reasonable to me. As @squaremo suggested, there will also be remote driver changes, but should not be major.
ping @mrjana

tomdee commented:
When I started writing the code I realized that my API isn't quite right. The static routes that the driver is asking libnetwork to add are scoped per interface and not per endpoint.

I see two options for adding this to JoinInfo. @mavenugo  or @mrjana which is your preference?

1) Add a new InterfaceRoutes type which is similar to InterfaceNames. It would return a list of InterfaceRouteInfo objects, which have a SetRoutes([]net.IPNet) method and a Id() method (similar to InterfaceNameInfo).

2) Or extend and rename InterfaceNames. It would become simply Interfaces. InterfacenameInfo would become InterfaceJoinInfo and would gain a SetRoutes([]net.IPNet) method.

Either way, I will also need a SetRoutesIPv6 method too.

tomdee commented:
I went for Option 1).  I've squashed, rebased and signed off my commit. What else do I need to do to get this reviewed and merged? 

I can make the changes to the remote driver once the code for that goes in.

tomdee commented:
ping @mavenugo I just wanted to check that you'd seen this PR and were happy with the general approach?

mrjana commented:
@tomdee A general comment. The data model for this should be in the `EndpointInfo/Sandbox` not the `Interface` as there is not reason to tie this to an interface. The kernel can always resolve it to an interface. If it can't it will fail.

tomdee commented:
@mrjana No, the point of this PR is to allow routes to be provided that the kernel _can't_ resolve to an interface. For Project Calico, the containers are assigned IPs on a /32 subnet - there is no gateway address on that network that the kernel can resolve (since there are no other addresses). We need to be able to explicitly tie a route to an interface so that traffic can be routed to our gateway over the interface.

We could use point-to-point /30 links but that would increase IP usage four fold. Allowing support for drivers to specify per-interface routes is the only way to allow us to continue assigning /32s

mrjana commented:
@tomdee Nevertheless, I think the api should still be generic and at the same time satisfy Project Calico requirements. So I suggest instead of only allowing interface specific routes we can still make routes to be nexthop based or interface based (either of them optional but one of them should be present) and for interface you can simply use the interface ID as the paramater so that inside libnetwork we can resolve that to the real interface inside the namespace.

spikecurtis commented:
@mrjana Tom's out of the office today, can I just check I've understood: is it still your preference that the data be under `EndpointInfo/Sandbox`

mrjana commented:
@spikecurtis Yes, but to satisfy Calico's requirement we can also pass the interface ID and if it is present we can resolve that to the right interface name inside the container and use it to install a interface specific static route. This way we can satisfy installing both kinds of static routes.

mavenugo commented:
@tomdee @spikecurtis did you get a chance to address @mrjana's comments ?
If you want this feature to participate in next docker release, we need to get this resolved asap.

Symmetric commented:
@mavenugo , @spikecurtis completed the code on Friday, but it's not yet tested -- I'm looking at this today. Would it help you if I updated the PR with Spike's changes so that you can take a look while I'm fixing the tests? Or would you prefer to wait until it's ready to merge?

mrjana commented:
@Symmetric Please push the existing untested changed. We can try to provide some early comments

Symmetric commented:
@mrjana I don't have admin access on the metaswitch/libnetwork fork, so I'm moving to a new repo/PR: please see https://github.com/docker/libnetwork/pull/216 for the latest.

If you want to merge under this PR (to preserve the record of discussion) we can fix that up first thing tomorrow.

Symmetric commented:
@mrjana I've just updated this PR with Spike's changes, and a first round of fixes to get the existing UTs passing again.

PTAL

mrjana commented:
@Symmetric Thanks for pushing this. One general piece that is missing is to remove the static routes when the container leaves the endpoint. Can you please add that to this PR? I will still continue to review the existing diffs.

mrjana commented on driverapi/driverapi.go:
```diff
@@ -104,6 +104,10 @@ type JoinInfo interface {
 
+	// SetStaticRoutes sets the routes that should be added to the sandbox.  It
+	// may be used in addtion to or instead of a default gateway (as above).
+	SetStaticRoutes(routes []types.StaticRoute) error
```
One common guideline that we have tried to maintain is to not expose structures in the interface. So instead of `SetStaticRoutes` accepting a type can we change it to `AddStaticRoute(routeType, Prefix, nh, interfaceID)` kind of an api which the driver can call multiple times to add as many routes? This will then work very similar to `AddInterface` api.

mrjana commented on endpoint_info.go:
```diff
@@ -149,6 +151,33 @@ func (ep *endpoint) InterfaceNames() []driverapi.InterfaceNameInfo {
+	return nil
+}
+
+func (ep *endpoint) AddInterfaceRoute(route *types.StaticRoute) error {
```
since this is not going to an exported api, can you make `AddInterfaceRoute` not exported?

mrjana commented on endpoint_info.go:
```diff
@@ -149,6 +151,33 @@ func (ep *endpoint) InterfaceNames() []driverapi.InterfaceNameInfo {
+		if r.RouteType == types.NEXTHOP {
+			ep.joinInfo.StaticRoutes = append(ep.joinInfo.StaticRoutes, &r)
+		} else {
+			if err := ep.AddInterfaceRoute(&r); err != nil {
```
If this fails we need to revert the `StaticRoutes` added to `joinInfo` but if we are going to change the interface to `AddStaticRoute` then this comment may be moot.

mrjana commented on sandbox/namespace_linux.go:
```diff
@@ -254,6 +255,14 @@ func (n *networkNamespace) SetGatewayIPv6(gw net.IP) error {
+func (n *networkNamespace) AddStaticRoute(r *types.StaticRoute) error {
+	err := programRoute(n.path, r.Destination, r.NextHop)
+	if err == nil {
+		n.sinfo.StaticRoutes = append(n.sinfo.StaticRoutes, r)
```
Please acquire the n.Lock() while appending to `n.sinfo.StaticRoutes`

mrjana commented on types/types.go:
```diff
@@ -184,6 +184,40 @@ func CompareIPNet(a, b *net.IPNet) bool {
+	NEXTHOP = iota
+
+	// CONNECTED indicates a StaticRoute with a interface for directly connected peers.
+	CONNECTED = iota
```
You don't have to initialize with `iota` again. It auto increments automatically

Symmetric commented:
@mrjana , thanks for the feedback -- I've applied your suggestions and pushed my latest commits. Any further suggestions, or should I squash-rebase now?

mrjana commented on endpoint.go:
```diff
@@ -360,6 +368,14 @@ func (ep *endpoint) Leave(containerID string, options ...EndpointOption) error {
+	for _, r := range ep.joinInfo.StaticRoutes {
+		err = sb.RemoveStaticRoute(r)
+		if err != nil {
+			return err
```
Even if this fails we want to just log an error message and continue and proceed forward with the rest of the leave.

mrjana commented on sandbox/namespace_linux.go:
```diff
@@ -254,6 +255,37 @@ func (n *networkNamespace) SetGatewayIPv6(gw net.IP) error {
+}
+
+func (n *networkNamespace) RemoveStaticRoute(r *types.StaticRoute) error {
+	n.Lock()
```
In libnetwork we never lock the code but only the data so you can just move this `Lock/Unlock` inside the `if`

mrjana commented on sandbox/namespace_linux.go:
```diff
@@ -254,6 +255,37 @@ func (n *networkNamespace) SetGatewayIPv6(gw net.IP) error {
 
+func (n *networkNamespace) AddStaticRoute(r *types.StaticRoute) error {
+	n.Lock()
+	defer n.Unlock()
```
In libnetwork we never lock the code only the data. Please move the `Lock/Unlock` inside the `if` clause

mrjana commented on endpoint.go:
```diff
@@ -360,6 +368,14 @@ func (ep *endpoint) Leave(containerID string, options ...EndpointOption) error {
+			return err
+		}
+	}
+
```
Also what about the interface routes? Do they get removed when the interfaces are removed or will that be an error in the kernel when we try to remove the interfaces when there are active routes referencing them?

mrjana commented on endpoint.go:
```diff
@@ -360,6 +368,14 @@ func (ep *endpoint) Leave(containerID string, options ...EndpointOption) error {
+		if err != nil {
+			logrus.Debugf("Remove route failed: %v", err)
+		}
+	}
```
I had a comment which got overwritten. What about interface routes? Would they get removed when we remove interfaces or does the kernel not allow interfaces to be removed when there are active routes referencing the interface?

Symmetric replied:
Not sure the precise mechanism you're using to detach the interface on Leave -- but if you delete an interface, its device routes go away. Also if you delete the namespace, the interfaces and routes are all cleared too. And if you move an interface between netns, the device routes are also removed. 

So I suspect it's fine to leave them hanging -- does that sound like it will be general enough?

mrjana replied:
Ok, I wasn't sure about what happens when interfaces move netns because that's what happens when you Leave. Yeah the routes will go away along with the interfaces when we destroy the namespace. I think this is general enough

mrjana commented:
@Symmetric Can you please squash the commits and rebase? And then I think we are good to merge this.

mrjana commented on sandbox/namespace_linux.go:
```diff
@@ -254,6 +255,35 @@ func (n *networkNamespace) SetGatewayIPv6(gw net.IP) error {
+		lastIndex := len(n.sinfo.StaticRoutes) - 1
+		for i, v := range n.sinfo.StaticRoutes {
+			if v == r {
+				n.Lock()
```
Well actually this should wrap around the whole code inside the `if` clause

Symmetric commented:
@mrjana I've fixed the last issue (locking in namespace_linux.go) and rebased. LMK if there's anything else you need. Thanks!

mrjana commented:
@Symmetric Thanks for taking care of all the comments real quick. The current diff LGTM. We just one more LGTM from another maintainer and then we can merge.

mavenugo commented:
@mrjana Thanks for the review. @Symmetric for the quick turnaround.
I will review it tonight.

tomdee commented:
Ping @mavenugo Did you get a chance to take a look at this last night?

lxpollitt commented:
@mavenugo @mrjana Can we get that one more LGTM so we can go ahead and merge this today?  (We need this API change for the Calico driver to work with Docker 1.7, so we are super keen to get it before the code freeze.)

mavenugo commented:
LGTM. Thanks for the patch.
